### DISCUSSION TITLE
Performance issue

### DISCUSSION THREAD

**Discussion by theotherraycole** [DistroAV/DistroAV#872] [General]
I found obs-ndi has been increasing the time to render frames about 8ms on two of my systems whenever I use it as a filter on a source. After some research, I found this greatly reduces it and my render times have gone from 10ms down to 2ms:

It appears void ndi_filter_offscreen_render(void *data, uint32_t cx, uint32_t cy) calls gs_stagesurface_map constantly when it is not necessary. I changed this section of obs-ndi-filter.cpp as follows. No idea if this is "correct", but it is working great for me and seems to be more efficient.

   	struct video_frame output_frame;
		if (video_output_lock_frame(s->video_output, &output_frame, 1,
					    os_gettime_ns())) {

// removed the unmap call and reset of video_data to NULL that was here...constant map/unmap seems to be a big hit
	
			gs_stage_texture(
				s->stagesurface,
				gs_texrender_get_texture(s->texrender));

// Only map once...mapping multiple times results in higher CPU
            if (s->video_data == NULL)
            {
    			gs_stagesurface_map(s->stagesurface, &s->video_data,
            					    &s->video_linesize);

    			gs_stagesurface_unmap(s->stagesurface);
            };

			uint32_t linesize = output_frame.linesize[0];
			for (uint32_t i = 0; i < s->known_height; ++i) {
				uint32_t dst_offset = linesize * i;
				uint32_t src_offset = s->video_linesize * i;
				memcpy(output_frame.data[0] + dst_offset,
				       s->video_data + src_offset, linesize);
			}

			video_output_unlock_frame(s->video_output);
		}

**Comment by paulpv** [comment]
Interesting. I'll give this a look.

**Reply by Trouffman** [reply]
@paulpv did you check to implement this in one the future release?

**Comment by loonietoque** [comment]
I've also run into this recently.

For context, it was on a system without a dedicated GPU - just an Intel 9700k with integrated graphics.

Enabling the Main NDI output doesn't result in much overhead (even in RGB mode instead of NV12, for what it's worth), but using a filter definitely does, to the point where it's not really usable for 60fps feeds.

**Comment by Trouffman** [comment]
This will be managed in : https://github.com/DistroAV/DistroAV/pull/1266
